sql/expression/function.go

```diff
@@ -65,3 +67,107 @@ func isBinary(data []byte) bool {
 
 	return true
 }
+
+// Substring is a function to return a part of a string.
+// This function follows the following rules:
+// - negative start or length causes an error.
+// - start bigger or equal to str length causes an error.
+// - valid start with 0 length returns the empty string.
+// - null string returns null.
+// Since Go strings are UTF8, this function does not return a direct sub
+// string str[start:start+length], instead returns the substring of rune
+// s. That is, "á"[0:1] does not return a partial unicode glyph, but "á"
+// itself.
+type Substring struct {
+	str   sql.Expression
+	start sql.Expression
+	len   sql.Expression
+}
+
+// NewSubstring creates a new substring UDF.
+func NewSubstring(str, start, len sql.Expression) sql.Expression {
+	return &Substring{str, start, len}
+}
+
+// Eval implements the Expression interface.
+func (s *Substring) Eval(row sql.Row) (interface{}, error) {
+	str, err := s.str.Eval(row)
+	if err != nil {
+		return nil, err
+	}
+
+	var text []rune
+	switch str := str.(type) {
+	case string:
+		text = []rune(str)
+	case []byte:
+		text = []rune(string(str))
+	case nil:
+		return nil, nil
+	default:
+		return nil, sql.ErrInvalidType.New(reflect.TypeOf(str).String())
+	}
+
+	start, err := s.start.Eval(row)
+	if err != nil {
+		return nil, err
+	}
+
+	start, err = sql.Int64.Convert(start)
+	if err != nil {
+		return nil, err
+	}
+
+	ln, err := s.len.Eval(row)
+	if err != nil {
+		return nil, err
+	}
+
+	ln, err = sql.Int64.Convert(ln)
+	if err != nil {
+		return nil, err
+	}
+
+	length := ln.(int64)
+	startIdx := start.(int64)
+	runeCount := int64(len(text))
+
+	if length < 0 {
+		return nil, fmt.Errorf("substring: invalid length %d", length)
+	}
+
+	if startIdx < 0 {
```

Comment:
Take into account that:
- The first position in string is 1.
- If start_position is a positive number, then the SUBSTRING function starts from the beginning of the string.
- If start_position is a negative number, then the SUBSTRING function starts from the end of the string and counts backwards. Negative values for start_position was introduced in MySQL 4.1.
- If start_position is bigger than the string length, the result is NULL
- If start_position is 0, the result is NULL

Reply:
I think the UI is wrong. Check with: `SELECT SUBSTRING("SQL Tutorial", 5, 3) AS ExtractString, NULL as test;` the output of `test` appears to be "" but it is NULL

Reply:
If you do `SELECT NULL is null` returns `1`, and if you do  `SELECT SUBSTRING("fooo bar", 0) is null` it returns `0`, so in this case is not null what it returns. From what I see in that UI, the only way SUBSTRING returns null is if the string itself is null.

Reply:
https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_substring

> If len is less than 1, the result is the empty string.